tools/validate/validate.cpp

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#define _CRT_SECURE_NO_WARNINGS
+#include <algorithm>
+#include <array>
+#include <assert.h>
+#include <exception>
+#include <filesystem>
+#include <stdexcept>
+#include <stdio.h>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+using namespace std;
+
+class BinaryFile {
+public:
+    explicit BinaryFile(const string& filepath) {
+        m_file = fopen(filepath.c_str(), "rb");
+
+        if (!m_file) {
+            throw runtime_error("fopen() failed for: " + filepath);
+        }
+    }
+
+    [[nodiscard]] bool read_next_block(vector<unsigned char>& buffer) {
+        constexpr size_t BlockSize = 65536;
+
+        buffer.resize(BlockSize);
+
+        const size_t bytes_read = fread(buffer.data(), 1, BlockSize, m_file);
+
+        buffer.resize(bytes_read);
+
+        return !buffer.empty();
+    }
+
+    ~BinaryFile() {
+        if (fclose(m_file) != 0) {
+            fprintf(stderr, "fclose() failed.\n");
+            terminate();
+        }
+    }
+
+    BinaryFile(const BinaryFile&) = delete;
+    BinaryFile& operator=(const BinaryFile&) = delete;
+
+private:
+    FILE* m_file{nullptr};
+};
+
+void scan_file(const string& filepath, vector<unsigned char>& buffer) {
+    constexpr char CR = '\r';
+    constexpr char LF = '\n';
+
+    bool has_cr       = false;
+    bool has_lf       = false;
+    bool has_crlf     = false;
+    bool has_utf8_bom = false;
+
+    size_t disallowed_characters     = 0;
+    size_t tab_characters            = 0;
+    size_t trailing_whitespace_lines = 0;
+
+    unsigned char prev      = '@';
+    unsigned char previous2 = '@';
+    unsigned char previous3 = '@';
+
+    for (BinaryFile binary_file{filepath}; binary_file.read_next_block(buffer);) {
+        for (const auto& ch : buffer) {
+            if (prev == CR) {
+                if (ch == LF) {
+                    has_crlf = true;
+                } else {
+                    has_cr = true;
+                }
+            } else {
+                if (ch == LF) {
+                    has_lf = true;
+                }
+            }
+
+            if (ch == '\t') {
+                ++tab_characters;
+            } else if (ch == 0xEF || ch == 0xBB || ch == 0xBF) {
+                has_utf8_bom = true;
+            } else if (ch != CR && ch != LF && !(ch >= 0x20 && ch <= 0x7E)) {

```

Comment:
can we have comments for what these character codes are ?

Reply:
I think the UTF-8 BOM one is self explanatory. 7E is a little self explanatory (7 bit clean), not sure about 20.

Reply:
I'll add comments and Wikipedia citations. Note that `0x7F` contains only 7 bits, yet is disallowed (it's the DEL control character).